lib/Inst/Inst.cpp

```diff
@@ -1174,16 +1174,68 @@ Inst *souper::getInstCopy(Inst *I, InstContext &IC,
   return Copy;
 }
 
+bool isTerminalInst(Inst *I) {
+  bool terminal;
+  switch (I->K) {
+    case Inst::Const:
+    case Inst::UntypedConst:
+    case Inst::Var:
+    case Inst::Phi:
+    case Inst::Hole:
+      terminal = true;
+      break;
+    default:
+      terminal = false;
+  }
+
+  return terminal;
+}
+
+// Return true if NewInst is equivalent to I
+bool isEquivalent(Inst *NewInst, Inst *I) {
+  // bfs on smaller instruction (NewInst)
+  if (I->K != NewInst->K || I->Ops.size() != NewInst->Ops.size())
+    return false;
+
+  if (isTerminalInst(I) && NewInst != I) {
+    return false;
+  }
+
+  for (int i = 0; i < I->Ops.size(); i++) {
+    if (!isEquivalent(I->Ops[i], NewInst->Ops[i]))
+      return false;
+  }
+
+  return true;
+}
+
+// Return true if NewInst is present in Cand.
+bool DAGMatch(Inst *NewInst, Inst *Cand) {
+  std::vector<Inst*> Results;
+  souper::findInsts(Cand, Results, [NewInst](Inst *I) {
+    if (isEquivalent(NewInst, I))
```

Comment:
just "return isEquivalent(NewInst, I)" the conditional is only confusing

Reply:
yeah, i agree. silly me.